File: Design-Patterns-Using-Ruby/Composite/CompositePattern.rb
# Component
class Component
    def operation
        raise NotImplementedError, 'Subclasses must implement this method'
    end
end

# Composite
class Composite < Component
    def initialize
        @children = []
    end
  
    def operation
        puts 'Composite Operation'
        @children.each(&:operation)
    end
  
    def add(component)
        @children << component
    end
  
    def remove(component)
        @children.delete(component)
    end
end

# Leaf
class Leaf < Component
    def operation
        puts 'Leaf Operation'
    end
end

# Client code
composite = Composite.new
composite.add(Leaf.new)

composite2 = Composite.new
composite2.add(Leaf.new)

composite.add(composite2)
composite.operation

=begin 
Composite Operation
Leaf Operation
Composite Operation
Leaf Operation
=end

File: Design-Patterns-Using-CSharp/Composite/CompositePattern.cs
using System;
using System.Collections.Generic;

// Component
abstract class Component
{
    public abstract void Operation();
}

// Composite
class Composite : Component
{
    private List<Component> children = new List<Component>();

    public override void Operation()
    {
        Console.WriteLine("Composite Operation");
        foreach (Component child in children)
        {
            child.Operation();
        }
    }

    public void Add(Component component)
    {
        children.Add(component);
    }

    public void Remove(Component component)
    {
        children.Remove(component);
    }
}

// Leaf
class Leaf : Component
{
    public override void Operation()
    {
        Console.WriteLine("Leaf Operation");
    }
}

// Client code
class CompositePattern
{
    static void Main(string[] args)
    {
        Composite composite = new Composite();
        composite.Add(new Leaf());

        Composite composite2 = new Composite();
        composite2.Add(new Leaf());

        composite.Add(composite2);
        composite.Operation();
    }
}

/*
Composite Operation
Leaf Operation
Composite Operation
Leaf Operation
*/